diff --git a/lld/MachO/UnwindInfoSection.cpp b/lld/MachO/UnwindInfoSection.cpp
index 2dbd8ed202728..7c2dd51430c19 100644
--- a/lld/MachO/UnwindInfoSection.cpp
+++ b/lld/MachO/UnwindInfoSection.cpp
@@ -343,7 +343,8 @@ void UnwindInfoSectionImpl::relocateCompactUnwind(
     if (!d->unwindEntry)
       return;

-    // If we have DWARF unwind info, create a CU entry that points to it.
+    // If we have DWARF unwind info, create a slimmed-down CU entry that points
+    // to it.
     if (d->unwindEntry->getName() == section_names::ehFrame) {
       // The unwinder will look for the DWARF entry starting at the hint,
       // assuming the hint points to a valid CFI record start. If it
@@ -360,7 +361,9 @@ void UnwindInfoSectionImpl::relocateCompactUnwind(
       cu.encoding = target->modeDwarfEncoding | dwarfOffsetHint;
       const FDE &fde = cast<ObjFile>(d->getFile())->fdes[d->unwindEntry];
       cu.functionLength = fde.funcLength;
-      cu.personality = fde.personality;
+      // Omit the DWARF personality from compact-unwind entry so that we
+      // don't need to encode it.
+      cu.personality = nullptr;
       cu.lsda = fde.lsda;
       return;
     }
