--- a/flang/include/flang/Runtime/command.h
+++ b/flang/include/flang/Runtime/command.h
@@ -14,7 +14,7 @@

 #ifdef _WIN32
 // On Windows* OS GetCurrentProcessId returns DWORD aka uint32_t
-typedef std::uint32_t pid_t;
+//typedef std::uint32_t pid_t;
 #else
 #include "sys/types.h" //pid_t
 #endif
--- a/flang/include/flang/Optimizer/Builder/Runtime/RTBuilder.h
+++ b/flang/include/flang/Optimizer/Builder/Runtime/RTBuilder.h
@@ -31,7 +31,7 @@

 #ifdef _WIN32
 // On Windows* OS GetCurrentProcessId returns DWORD aka uint32_t
-typedef std::uint32_t pid_t;
+//typedef std::uint32_t pid_t;
 #endif

 // Incomplete type indicating C99 complex ABI in interfaces. Beware, _Complex
